chess_env.py

```diff
@@ -11,20 +16,120 @@ def _state_to_uci(self):
     def _uci_to_state(self):
         pass
 
-    def _push_move(self):
-        pass
+    # takes tuple (x,y) and converts to uci repr
+    # Eg - (1,2) = 'a2'
+    def _pos_to_uci(self, pos):
+        x, y = pos # x, y are in the range 1 to 8
+        uci_x = chr(x + 96) # 97 = 'a'
+        uci_y = str(y)
+        return uci_x + uci_y
+
+    # takes uci representation for one position
+    # and converts to (x,y)
+    # Eg - 'a2' = (1,2)
+    def _uci_to_pos(self, uci):
+        uci_x, uci_y = uci
+        x, y = ord(uci_x) - 96, int(uci_y) # 'a' = 97
+        return (x,y)
+
+    def _action_index_from_uci(self, move_uci):
+        piece_map = self.board.piece_map()
+        initial_pos = self._uci_to_pos(move_uci[0:2])
+        piece = piece_map[ initial_pos[0]*8 + initial_pos[y] ]
+        if piece.color != self.player_color:
+            return -1
+        next_pos = self._uci_to_pos(move_uci[2:])
+        x,y = initial_pos
+        nx, ny = next_pos
+        dx, dy = nx - x, ny - y
+        if str.lower(piece.symbol()) == 'k': # king move
+            king_delta_action_dict = { (-1,1) : 0, (0, 1): 1, (1, 1): 2, (1, 0): 3, \
+                                                        (1, -1): 4, (0, -1): 5, (-1, -1):6, (-1, 0): 7}
+            action = king_delta_action_dict[(dx, dy)]
+        elif str.lower(piece.symbol()) == 'r':
+            if dx == 0: # vertical move
+                if dy < 0: # down
+                    action = 8 + 15 + abs(dy) # 8 - king, 15 - horizontal
+                else:
+                    action = 8 + 15 + 15 + abs(dy) # 8 - king, 15 - horizontal, 7 - down
+            else: # horizontal move
+                if dx < 0: # left
+                    action = 8 + abs(dx) # 8 - king
+                else:
+                    action = 8 + 7 + abs(dx) # 8 - king
+        return action
+
+    # action is of the form -
+    # [Kdiagupleft, Kup, Kdiagupright,
+    #   Kright,
+    #   Kdiagdownright, Kdown, Kdiagdownleft,
+    #   Kleft,
+    #   RookHorizontal: -7 to + 7 (excluding 0),
+    #   RookVertical: -7 to +7 (excluding 0)]
+    # Total dims - 8 + 14 + 14 = 36
+    # Thus, input here is from 0 to 35 #TODO: correct
+    # [K1..K8, -7..-1, 1..7, -7..-1,1..7]
+    # [0...8, 9...15, 16...22, 23...29, 30...36]
```

Comment:
King moves from 0 to 7, not 0 to 8. All other indices should be changed accordingly.